IntegrationTests/UITests/UITest.cs

```diff
@@ -1,58 +1,181 @@
 ﻿using System;
+using System.Diagnostics;
 using System.Linq;
+using System.Reflection;
 using System.Threading.Tasks;
 using System.Windows.Forms;
+using CommonTestUtils;
 using GitUI;
+using NUnit.Framework;
 
 namespace GitExtensions.UITests
 {
     public static class UITest
     {
-        public static async Task WaitForIdleAsync()
+        public static void RunForm<T>(
+            Action showForm,
+            Func<T, Task> runTestAsync)
+            where T : Form
         {
+            LoggingService.Log($"{MethodBase.GetCurrentMethod().Name} entry");
             var idleCompletionSource = new TaskCompletionSource<VoidResult>();
+
+            // Start runAsync before calling showForm.
+            // The latter might block until the form is closed, especially if using Application.Run(form).
+
+            // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
+            // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
+            // complete.
+            var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
+            {
+                LoggingService.Log($"{MethodBase.GetCurrentMethod().Name}.test entry");
+
+                // Wait for the form to be opened by the test thread.
+                await idleCompletionSource.Task;
+                LoggingService.Log($"{MethodBase.GetCurrentMethod().Name}.test idle");
+                var form = Application.OpenForms.OfType<T>().Single();
+                if (!form.Visible)
+                {
+                    var errorMessage = $"{form.GetType().FullName} {form.Name} should be visible";
+                    form.Dispose();
+                    Assert.Fail(errorMessage);
+                }
+
+                try
+                {
+                    // Wait for potential pending asynchronous tasks triggered by the form.
+                    WaitForPendingOperations();
```

Comment:
❗️ This should not be part of the test initialization. UI tests wait for the UI operation to complete, but *do not* wait for background operations to complete. If a test depends on background operations completing, the test needs to include that wait as part of the test itself.

Reply:
This may apply for the generic theory of testing.
For GE, which uses asynchronous tasks extensively, we should wait for the background operations by default in order to have stable tests. This little additional runtime will not hurt.
We can add an option to switch back to `await WaitForIdleAsync();` for tests that explicitely want to start before the background tasks have finished.

Reply:
> For GE, which uses asynchronous tasks extensively, we should wait for the background operations by default in order to have stable tests.

We don't need to wait for the operations by default to have stable tests. Most of the tests I wrote had no dependency on operations except for ones tied directly to the UI thread, so waiting was simply unnecessary overhead. For the others, adding a wait as part of the test improves the clarity of the test itself (showing that a specific test covers UI state that might differ from what a user sees when the form first opens).

Reply:
I agree, that it would improve the clarity of the test cases. But I will not review every single UI test. This little overhead will not hurt.

Reply:
You only need to review the tests that have been known to fail.